### PR TITLE
Labs spring web

### PR DESCRIPTION
**Pull request by LiveNathan**
Hey Nikita, thanks for taking a look at this. Most of it is the code copied directly from Ryan. Hopefully you can see the newer things I added. Here are some questions that came  up for me that we can discuss here or during our meeting tomorrow.
- [ ] How to provide a custom message to the user with @NotNull since their request will be rejected.
- [ ] What should I do about Intellij's warning about using @Data with JPA entities? Follow it? Ignore it?

### CODE REVIEW

**Comment by LiveNathan** [comment]
@NikitaBaranov I decided to go ahead and merge this PR in order to continue with the Testing Endpoints Using jsonPath()
 lesson.

File: src/main/java/cn/RecipeAPI/Services/ReviewService.java [hunk 1]
```diff
@@ -45,13 +48,30 @@ public List<Review> getReviewByUsername(String username) throws NoSuchReviewExce
...
+        if (review.getRating() == null) {
+            throw new NoEmptyRatingException("This review needs a rating!");
+        }
         recipe.getReviews().add(review);
+        recipe.setAverageRating(calculateAverageRecipeRating(recipeId));  // Recalculate average every time a new review is added.
         recipeService.updateRecipe(recipe, false);
         return recipe;
     }
 
+    public int calculateAverageRecipeRating(Long recipeId) throws NoSuchReviewException, NoSuchRecipeException {
```
Comment (NikitaBaranov) [inline src/main/java/cn/RecipeAPI/Services/ReviewService.java:65]: I believe you need to do the same calculation on Update and Delete :)

File: src/main/java/cn/RecipeAPI/Models/Review.java [hunk 2]
```diff
@@ -23,16 +23,17 @@ public class Review {
     @NotNull
     private String username;
 
-    private int rating;
+    private Integer rating;
```
Comment (NikitaBaranov) [inline src/main/java/cn/RecipeAPI/Models/Review.java:26]: Since it is an object now it can be `null` so it could be a good idea to check it on line 32

**Review by NikitaBaranov — COMMENTED** [review]
Great code.
I left a couple of suggestions for you to implement in your next PR

**Comment by NikitaBaranov** [comment]
> @NikitaBaranov I decided to go ahead and merge this PR in order to continue with the Testing Endpoints Using jsonPath() lesson.

Thank you!  I'm sorry I was a bit busy :( 

In general, please definitely go ahead once you have the next task, don't be blocked.

File: src/main/java/cn/RecipeAPI/Services/ReviewService.java (hunk 1 above)
Comment (LiveNathan) [reply src/main/java/cn/RecipeAPI/Services/ReviewService.java:65]: Ok, I think I fixed it and it should show up in a future PR. This is the query method I wrote that I *hope* will work:
`Recipe findByReviewContains(Review review);`

**Pull request merged**